documentation: boardfarm tests integration design review

abelog opened the pull request:
## Description

According to #1116 , every new feature needs a design document.
Migration from `test_flows.py` to Boardfarm for HW testing was discussed and agreed on before #1116. But this activity has to be documented in any case, to avoid effort which we had to apply for documenting `test_flows.py`

This PR proposes documentation which describes the end state of boardfarm integration.
Idea is to modify it while development is going on if something is changed.

Related to #1074 .

abelog commented:
Small **note**:
Seems, that images are not rendered properly, as Github looks for them in master branch. But those images are introduced by this PR.

abelog committed:
documentation: add classes UML image for boardfarm tests integration

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

abelog committed:
documentation: add typical setup image for boardfarm tests integration

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog committed:
documentation: design review for boardfarm integration in prplMesh

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

arnout commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,85 @@
+[boardfarm](https://github.com/mattsm/boardfarm/) already has some representation of some general devices (like Linux, OpenWRT device).
+But our test flows involve more components and require more info about devices\[1\].
+So we have to expand base classes with representation of `prplWRT` devices capable handling Agent, Controller and Sniffer.
+Also, some additional abstraction of easyMesh specific entities may ease future work(like CAPI, IEEE1905.1 message).
```
```suggestion
Also, some additional abstraction of easyMesh specific entities may ease future work (like CAPI, IEEE1905.1 message).
```

abelog replied:
Fixed.

arnout commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,85 @@
+	1. prplMesh_controller
+	1. prplMesh_agent
+	1. prplMesh_radio_dummy
+	1. prplMesh_radio_hw
```
It would be better to use the actual names that we end up using:
```suggestion
	1. UCCSocket
	1. ALEntity
        1. Radio
            1. RadioDocker
            1. RadioHostapd
        1. VirtualAP
            1. VirtualAPDocker
            1. VirtualAPHostapd
```

Bonus points for updating the UML files with these names as well :-)

abelog replied:
I Have added a new commit with UML update. Let me know if got it right.

arnout commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,85 @@
+[boardfarm](https://github.com/mattsm/boardfarm/) itself supports plugins\[2\].
+It allows to embed code to [boardfarm](https://github.com/mattsm/boardfarm/) which isn't present in main repository.
+All device classes, test specific entities and tests itself should be included like this.
+When they are stable enough, it's expected to push that to original [boardfarm](https://github.com/mattsm/boardfarm/) repository.
```
Actually, I don't think this is true. The classes we add are only useful for prplmesh-specific tests, and the devices we add are only useful to make the link.

abelog replied:
Then i will remove that mention. Having at least a link to those files may help other who seeks for examples of new devices.

arnout commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,85 @@
+1. Create infrastructure.
+1. Move all tests from `test_flows.py`.
+1. Form boardfarm configuration for CI set up.
+1. Set up GitLab CI to trigger tests on PR merge.
```
I don't think this is relevant in an architecture document that sits in the repository.

Instead, refer to /tests/boardfarm_plugins/README.md

abelog replied:
Ok. Removing it.

arnout reviewed (approved):
Small things only.

abelog committed:
documentation: boardfarm_prplMesh_classes.puml: update to reality

Update boardfarm_prplMesh_classes.puml to correspond to what is actually
implemented or going to be implemented.

Add ALEntity, Radio, VirtualAP entities.

Rename prplmesh_docker to PrplMeshDocker, and add PrplMeshBase.

Remove member that don't exist.

Update relationships to reality. Also add some placement hints to them.

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented on documentation/images/plantuml/boardfarm_prplMesh_classes.puml:
```diff
@@ -96,7 +96,7 @@ package prplMesh_entities <<Frame>> {
         get_rxpower()
```
All of these don't exist at the moment. get_mac() will never exist - mac is accessed directly. status I'm not convinced if we will add it. tx and rxpower we may add in the future. So I propose to remove status and get_mac(). Also, mark the other functions as abstract please.

arnout commented on documentation/images/plantuml/boardfarm_prplMesh_classes.puml:
```diff
@@ -86,7 +86,7 @@ package prplMesh_entities <<Frame>> {
         check_log()
```
This is called wait_for_log instead of check_log. Also, it's an abstract function.

For radios, there is no accessor function. The relation with the Radio class is sufficient.

Also the class is ALEntity, not prplmesh_agent.

arnout reviewed (approved):
The last commit is still not entirely correct. Instead of explaining it all, I'm just going to fix it myself. Therefore, I'm approving and reassigning to myself.

tomereli commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,80 @@
+# Design
+
+By itself, [boardfarm](https://github.com/mattsm/boardfarm/) provides us with:
+1. Abstraction of test set-up
```
```suggestion
1. Abstraction of test setup
```

abelog replied:
Fixed!

tomereli commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,80 @@
+
+By itself, [boardfarm](https://github.com/mattsm/boardfarm/) provides us with:
+1. Abstraction of test set-up
+1. Provides some basic classes of devices
```
You already use "provides".
```suggestion
1. Basic device abstraction classes
```

abelog replied:
Applied.

tomereli commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,80 @@
+By itself, [boardfarm](https://github.com/mattsm/boardfarm/) provides us with:
+1. Abstraction of test set-up
+1. Provides some basic classes of devices
+1. Some simple tests
```
Such as? Please add an example if you're mentioning tests

abelog replied:
Those are generic tests, like: check memory usage, check CPU usage and etc. They are not much of use to us. So i will remove that mention.

tomereli commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,80 @@
+1. Abstraction of test set-up
+1. Provides some basic classes of devices
+1. Some simple tests
+1. Handles connections to boards
```
```suggestion
1. DUT connectivity (ssh, serial, telnet, etc.)
```

abelog replied:
Applied.

tomereli commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,80 @@
+1. Provides some basic classes of devices
+1. Some simple tests
+1. Handles connections to boards
+1. Provides test-suite organisation
```
```suggestion
1. Test-suite organization
```
But please, explain what this means. If you need to say a sentence about the YAML file describing a test-suite, than do so.

abelog replied:
It  allows us to have logical grouping of tests and bulk execution of them. As result, all reports and logs are formed for whole test suite (still, logs are also available per test so you don't have to look general log for specific failed test). We can also execute each test individually, but it's not convenient.
It's not is YML, it's just a plain text file with some formatting. [Example of such `.cfg` file in boardfarm repository](https://github.com/mattsm/boardfarm/blob/master/boardfarm/testsuites.cfg).

tomereli commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,80 @@
+
+[boardfarm](https://github.com/mattsm/boardfarm/) already has some representation of some general devices (like Linux, OpenWRT device).
+But our test flows involve more components and require more info about devices\[1\].
+So we have to expand base classes with representation of `prplWRT` devices capable handling Agent, Controller and Sniffer.
```
This sentence is unclear. I would expect to expand the base classes with EasyMesh device, which can be an agent or controller, so:
```
We expand base classes with the representation of `EasyMesh` devices capable of running as Controller, Agent, or both.
```
What does the sniffer has to do with it?

abelog replied:
Sniffer is not EasyMesh entity, it's standalone docker image/device in setup. Order of words has confused.
Moved Sniffer mention before the mention of EasyMesh-specific entities, to avoid confusion that Sniffer has anything to do with EasyMesh specific entities.

tomereli commented on documentation/prplMesh-test-boardfarm.md:
```diff
@@ -0,0 +1,80 @@
+![test setup](/documentation/images/plantuml/boardfarm_prplMesh_test_setup.png)
+
+Please, note.
+In general case, it's expected that prplMesh executables are delivered to DUT (device under test) as OpenWRT `*.ipk`.
```
Better way to say this IMO:
Note - normally, it is expected that the prplmehs IPK is deployed to the DUT (Device Under Test), or that the DUT is already flashed with prplmesh embedded in the image.
For this reason, LAN connection for flashing in uboot is not required and is deliberately missing from the scheme.

abelog replied:
Taken in with some fixes of mine.

tomereli reviewed (approved):
Mainly comments about the README.md file, but nothing critical so approving.

tomereli commented:
Approved but added the don't merge tag in case @abelog you want to fix my comments.

abelog commented:
@tomereli, i have applied suggestions of yours( see 7e4e7e1c698b1ef09c5fc09feb3966b75546762e). Please approve again if it's fine by you. Only after that i will rebase and merge  this document.

tomereli commented:
> @tomereli, i have applied suggestions of yours( see 7e4e7e1c698b1ef09c5fc09feb3966b75546762e). Please approve again if it's fine by you. Only after that i will rebase and merge  this document.

I already approved, you can merge

rmelotte commented:
@mergifyio rebase
